frontend/src/features/admin-form/common/mutations.ts

```diff
@@ -55,6 +64,72 @@ export const useMutateCollaborators = () => {
     [formId, queryClient],
   )
 
+  const getMappedBadRequestErrorMessage = (
+    formCollaboratorAction: FormCollaboratorAction,
+  ) => {
+    let badRequestErrorMessage
+    switch (formCollaboratorAction) {
+      case FormCollaboratorAction.ADD:
+        badRequestErrorMessage = `Please ensure that the email entered is a valid government email. If the error still persists, refresh and try again later.`
+        break
+      default:
+        badRequestErrorMessage = `Sorry, an error occurred. Please refresh the page and try again later.`
+    }
+
+    return badRequestErrorMessage
+  }
+
+  const getMappedDefaultErrorMessage = (
+    formCollaboratorAction: FormCollaboratorAction,
+  ) => {
+    let defaultErrorMessage
+    switch (formCollaboratorAction) {
+      case FormCollaboratorAction.ADD:
+        defaultErrorMessage = 'Error adding collaborator.'
+        break
+      case FormCollaboratorAction.UPDATE:
+        defaultErrorMessage = 'Error updating collaborator.'
+        break
+      case FormCollaboratorAction.REMOVE:
+        defaultErrorMessage = 'Error removing collaborator.'
+        break
+      case FormCollaboratorAction.REMOVE_SELF:
+        defaultErrorMessage = 'Error removing self.'
+        break
+      case FormCollaboratorAction.TRANSFER_OWNERSHIP:
+        defaultErrorMessage = 'Error transfering form ownership.'
+        break
+      //should not reach
+      default:
+        defaultErrorMessage = 'Error.'
+    }
+    return defaultErrorMessage
+  }
+
+  const getMappedErrorMessage = (
+    error: HttpError | Error,
+    formCollaboratorAction: FormCollaboratorAction,
+    requestEmail?: string,
+  ) => {
+    // check if error is an instance of HttpError to be able to access status code of error
+    if (error instanceof HttpError) {
+      let errorMessage
+      switch (error.code) {
+        case 422:
+          errorMessage = `${!requestEmail} is not part of a whitelisted agency`
```

Comment:
The `!` a not a non-null assertion, it’s the [logical not operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_NOT). It will convert whatever truthiness `requestEmail` represents to its opposite `true` or `false`, and so the resulting message will be either :
```
true is not part of a whitelisted agency
```
or
```
false is not part of a whitelisted agency
```
which is not what we want 😅 .

If `requestEmail` is always defined on 422 errors, we can make it clear with a cast:
errorMessage = `${requestEmail as string} is not part of a whitelisted agency`

If `requestEmail` may be `undefined`, we’d be better off with a conditional block:
```javascript
errorMessage = requestEmail
    ? `${requestEmail} is not part of a whitelisted agency`
    :  `An unexpected error 422 happened`
```

One small issue I have with doing a straight cast in this case is code self-consistency and implicit assumptions. Maybe we can try to make things a little clearer 🤔 . Based on the call tree of `getMappedErrorMessage()`,  `requestEmail` is only supplied if `formCollaboratorAction` is `FormCollaboratorAction.ADD`. The `422` error code is not inspected in a context where `formCollaboratorAction` has been checked, so if we just cast `requestEmail as string` to show we know it's true, we've implicitly stated **only** `FormCollaboratorAction.ADD` returns 422. It may be true, but it's a bit dirty and not obvious from the code itself (the function signature shows `requestEmail` can be `undefined`).

Long story short: I'd recommend going the defensive programming route here. Either do a check to verify `requestEmail` is not `undefined` like above, and have an error message for unexpected cases. Or check on  `formCollaboratorAction` first, and then cast to string, and you likely still need an error message for unexpected cases.

Reply:
Thanks for the detailed explanation. I initially assumed that it was okay to add an empty string if in the case `requestEmail` is empty because the 422 error thrown only occurs for `FormCollaboratorAction.Add`. I do realise now that it is not clear in the code that that is the case and it is better just to do a simple non-null check and throw a different error if `requestEmail` is null.